### DISCUSSION TITLE
Howto sign a Wineskin winery MacOS generated application package?

### DISCUSSION THREAD

**Discussion by RedondoARG77** [The-Wineskin-Project/WineskinServer#451] [Q&A] [answered]
We have a windows application where we create/generate via Wineskin Winery on MACOS, a MacOS compliant application package from.
So double clicking the Winery Application package just start the windows application correctly on MacOS. Everything fine till there.
FYI, we use the following environment:
 **MAC**: _Macbook Pro M1 (2020)_
 **Mac OS**: _Ventura 13.5_
 **Wine Version**: _7.7_
 **Engine Version**: _WS11WineCX64Bit22.1.1-rc1_
 **Wrapper Version**: _Wineskin-2.9.1.8_

We zip this Winery Application Package and in this form we want to distribute it to our customers and this is where the problem arises.
When we deliver it via WeTransfer or our own company Dropbox clone, gatekeeper jumps in when the customer downloads, unzip and launches the package. The message **_"<YOUR APP NAME> is damaged and can't be opened. You should move it to the Bin"._**
I've read that you can disable gatekeeper via some terminal command to allow the opening of the package, but this is not really what we want.
So, I tried to sign this package via the terminal commando 'codesign', but I'm not able to do so (we have an official Apple Developer Application certificate).

`codesign --sign "Developer ID Application certificate" --deep --force /Applications/<YOURAPPNAME>.app` returns the following error: _"unsealed contents present in bundle root"_

When I try to verify the Application bundle to see where the problem resides I get another error
`codesign --deep --verify --strict --verbose=2 /Applications/<YOURAPPNAME>.app` returns _"code has no resources but signature indicates they must be present."_

It seems that the package structure generated by Winery is not compliant with the Apple expected guidelines.
Is there a workaround or method to sign my package correctly?

**Comment by Gcenx** [comment]
I believe you need to remove `Wineskin.app` and also remove the symlinks that are placed next to it.

Outside of this I really can’t remember.

**Reply by RedondoARG77** [reply]
Ok... will give it a try.
The question that arises me is, if the signing succeeds after this action, if the application package still works like before or do I need to put the file and symlinks back after signing?

**Reply by Gcenx** [reply]
The symlinks and `Wineskin.app` are not required to run a wrapper when the main executable has been set.

I’ll eventually get around to changing this so these steps are not required, most of my time is spent rebasing wine fixes/hacks/workaround and library related issues.

**Reply by RedondoARG77** [reply]
I tried, before signing, to remove the symlinks and wineskin.app from the highest level (only the contents folder resides there now) but still the same error during signing the package: `"unsealed contents present in bundle root"`

Verifying the application bundle gives also again the error `code has no resources but signature indicates they must be present.`

Someone any other ideas?

**Reply by kimnzl** [reply, marked as answer]
One thing to check for is dot files in the root. I had one called ".wineskin_manager"
`ls -la ${appName}.app/`

I also had to correct the Info.plist
`plutil -replace CFBundleExecutable -string wineskinlauncher ${appName}.app/Contents/Info.plist`
`<key>CFBundleExecutable</key>
<string>WineskinLauncher</string>`
to
`<key>CFBundleExecutable</key>
<string>wineskinlauncher</string>`

**Reply by RedondoARG77** [reply]
Thank you all!
.
The combination of all your tips did the job.

Meaning:
- Removing all files/folders on the root level of the Wineskin package (except the 'Contents'-folder ;-)
- Remove the hidden file '_.wineskin_manager_'
- Modify the CFBundleExecutable key value in the Info.plist file. This can be easily done from the terminal with the command:
           `plutil -replace CFBundleExecutable -string wineskinlauncher ${appName}.app/Contents/Info.plist`

Verifying the signing of the package:
`codesign --verify --verbose <path to app>`

Gives this output:
**_< path to app >: valid on disk_**
**_< path to app >: satisfies its Designated Requirement_**
